File: pivotradar/platform/pivotradar_platform/calibration.py
"""Carga de calibración por activo desde JSON."""

from __future__ import annotations

import json
from dataclasses import dataclass, asdict
from pathlib import Path


@dataclass
class Calibration:
    symbol: str = "EURUSD"
    utc_offset_hours: int = 0
    point: float = 1e-5
    digits: int = 5
    ntfy_topic: str = ""
    ntfy_server: str = "https://ntfy.sh"
    engine: str = "rust"  # rust | reference

    n_ruptura: int = 4
    d1_atr_threshold: float = 0.5
    body_ratio_min: float = 0.4
    d1_use_retest: bool = True
    d1_use_volume: bool = True
    d1_min_volume: float = 1.2

    sweep_n: int = 6
    sweep_wick_min: float = 0.55
    reclaim_body_min: float = 0.55
    equal_hl_window: int = 10
    equal_hl_tol: float = 0.15
    d2_anticipar: bool = True

    fvg_min_size_atr: float = 0.2
    fvg_body_ratio: float = 0.55
    fvg_mitig_umbral: float = 0.5

    ob_lookback: int = 12
    ob_body_min: float = 0.4
    ob_impulse_min: float = 0.7

    mss_lookback_h4: int = 20
    mss_max_age_h4_bars: int = 12

    pivot_depth: int = 2
    pivot_lookback: int = 24
    sweep_distancia: float = 1.5
    zona_margen: float = 0.5
    peso_estructural: float = 0.25

    prob_d1: int = 65
    prob_d2: int = 70
    prob_d3: int = 65
    prob_d4: int = 65
    prob_d5: int = 75

    def prob_base_for(self, detector: str) -> int:
        if detector == "D1":
            return self.prob_d1
        if detector in ("D2", "D2_ANTICIPACION"):
            return self.prob_d2
        if detector in ("D3", "D3_DEF"):
            return self.prob_d3
        if detector == "D4":
            return self.prob_d4
        if detector == "D5":
            return self.prob_d5
        return 55

    def to_json(self) -> str:
        d = asdict(self)
        return json.dumps(d)

    @classmethod
    def from_dict(cls, d: dict) -> "Calibration":
        known = {f.name for f in cls.__dataclass_fields__.values()}
        return cls(**{k: v for k, v in d.items() if k in known})


def load_calibrations(path: str | Path | None = None) -> dict[str, Calibration]:
    """Carga `assets` de calibrations.json -> {symbol: Calibration}."""
    if path is None:
        path = Path(__file__).resolve().parent.parent / "config" / "calibrations.json"
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
    return {sym: Calibration.from_dict(cfg) for sym, cfg in data["assets"].items()}

File: pivotradar/rust/crates/core/src/calibration.rs
use serde::{Deserialize, Serialize};

/// Calibración por activo. Espejo de los `input` + `prob_base` del EA MQL5.
/// Todos los campos tienen default = valores EURUSD (referencia de calibración conservadora).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(default)]
pub struct Calibration {
    pub symbol: String,
    /// Offset en horas del servidor respecto a UTC (para sesiones/kill zones).
    pub utc_offset_hours: i32,
    pub ntfy_topic: String,
    pub ntfy_server: String,

    // --- D1 (Ruptura de rango)
    pub n_ruptura: i32,
    pub d1_atr_threshold: f64,
    pub body_ratio_min: f64,
    pub d1_use_retest: bool,
    pub d1_use_volume: bool,
    pub d1_min_volume: f64,

    // --- D2 (Liquidity Sweep + Reclaim)
    pub sweep_n: i32,
    pub sweep_wick_min: f64,
    pub reclaim_body_min: f64,
    pub equal_hl_window: i32,
    pub equal_hl_tol: f64,
    pub d2_anticipar: bool,

    // --- D3 (Fair Value Gap)
    pub fvg_min_size_atr: f64,
    pub fvg_body_ratio: f64,
    pub fvg_mitig_umbral: f64,

    // --- D4 (Order Block)
    pub ob_lookback: i32,
    pub ob_body_min: f64,
    pub ob_impulse_min: f64,

    // --- D5 (MSS H4 + Sweep)
    pub mss_lookback_h4: i32,
    pub mss_max_age_h4_bars: i32,

    // --- D0 (Estructura)
    pub pivot_depth: i32,
    pub pivot_lookback: i32,
    pub sweep_distancia: f64,
    pub zona_margen: f64,
    pub peso_estructural: f64,

    // --- Probabilidad base por detector (difiere por activo)
    pub prob_d1: i32,
    pub prob_d2: i32,
    pub prob_d3: i32,
    pub prob_d4: i32,
    pub prob_d5: i32,
}

impl Default for Calibration {
    fn default() -> Self {
        Calibration {
            symbol: String::from("EURUSD"),
            utc_offset_hours: 0,
            ntfy_topic: String::new(),
            ntfy_server: String::from("https://ntfy.sh"),
            n_ruptura: 4,
            d1_atr_threshold: 0.5,
            body_ratio_min: 0.4,
            d1_use_retest: true,
            d1_use_volume: true,
            d1_min_volume: 1.2,
            sweep_n: 6,
            sweep_wick_min: 0.55,
            reclaim_body_min: 0.55,
            equal_hl_window: 10,
            equal_hl_tol: 0.15,
            d2_anticipar: true,
            fvg_min_size_atr: 0.2,
            fvg_body_ratio: 0.55,
            fvg_mitig_umbral: 0.5,
            ob_lookback: 12,
            ob_body_min: 0.4,
            ob_impulse_min: 0.7,
            mss_lookback_h4: 20,
            mss_max_age_h4_bars: 12,
            pivot_depth: 2,
            pivot_lookback: 24,
            sweep_distancia: 1.5,
            zona_margen: 0.5,
            peso_estructural: 0.25,
            prob_d1: 65,
            prob_d2: 70,
            prob_d3: 65,
            prob_d4: 65,
            prob_d5: 75,
        }
    }
}

impl Calibration {
    /// Probabilidad base según detector (calibración por activo).
    pub fn prob_base_for(&self, det: &str) -> i32 {
        match det {
            "D1" => self.prob_d1,
            "D2" | "D2_ANTICIPACION" => self.prob_d2,
            "D3" | "D3_DEF" => self.prob_d3,
            "D4" => self.prob_d4,
            "D5" => self.prob_d5,
            _ => 55,
        }
    }
}
